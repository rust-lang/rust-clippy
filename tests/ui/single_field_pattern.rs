#![warn(clippy::single_field_patterns)]
struct Struct {
    field1: Option<i32>,
    field2: Option<i32>,
}

fn lint_struct(struct1: Struct) {
    let Struct { field1, .. } = struct1;
    let Struct { field1, field2: _ } = struct1;
    match struct1 {
        Struct { field1: Some(n), .. } if n >= 50 => {},
        Struct { field1: None, .. } => {},
        _ => {},
    }
    match struct1 {
        Struct { field1: Some(n), .. } if n >= 50 => {},
        Struct { .. } => {},
        _ => {},
    }
    match struct1 {
        Struct { field1: Some(1), .. } => {},
        Struct { field1: Some(2), .. } => {},
        Struct { field1: None, .. } => {},
        _ => {},
    }
    match struct1 {
        Struct {
            field1: Some(_) | None, ..
        } => {},
    }
    while let Struct { field1: Some(5), .. } = struct1 {}
}

struct Tuple(Option<i32>, Option<i32>);

fn lint_tuple_struct(tuple: Tuple) {
    if let Tuple(Some(x), _) = tuple {}
    if let Tuple(_, Some(y)) = tuple {}
    match tuple {
        Tuple(Some(1), ..) => {},
        Tuple(Some(2), ..) => {},
        Tuple(a, ..) => {},
    }
    match tuple {
        Tuple(Some(1) | Some(42) | Some(6082), ..) => {},
        Tuple(a, ..) => {},
    }
}

fn lint_tuple(tuple: (Option<i32>, Option<i32>)) {
    if let (Some(z), _) = tuple {}
    if let (_, Some(n)) = tuple {}
    match tuple {
        (Some(1), ..) => {},
        (Some(2), ..) => {},
        (a, ..) => {},
    }
    match tuple {
        (Some(1) | Some(42) | Some(6082), ..) => {},
        (a, ..) => {},
    }
}

fn lint_array(array: [i32; 3]) {
    match array {
        [1 | 2, ..] => {},
        [x @ 3, ..] => {},
        [r @ 20..=65, ..] => {},
        [e, ..] => {},
    }
    if let [5, ..] = array {}
}

fn ok_struct(struct1: Struct) {
    let _ = struct1;
    let Struct { field1, field2, .. } = struct1;
    let Struct { field1, field2 } = struct1;
    match struct1 {
        Struct {
            field1: Some(1),
            field2: _,
        } => {},
        Struct {
            field1: _,
            field2: Some(1),
        } => {},
        _ => {},
    }
    match struct1 {
        Struct { field1: Some(1), .. } => {},
        Struct {
            field1: _,
            field2: Some(1),
        } => {},
        _ => {},
    }
    let s @ Struct { field1, .. } = struct1;
}

fn ok_tuple_struct(tuple: Tuple) {
    if let Tuple(Some(1), two) = tuple {}
    if let Tuple(one, Some(1)) = tuple {}
    match tuple {
        Tuple(Some(1), _) => {},
        Tuple(_, Some(1)) => {},
        _ => {},
    }
    match tuple {
        Tuple(Some(1), ..) => {},
        Tuple(.., Some(1)) => {},
        _ => {},
    }
    match tuple {
        Tuple(.., Some(1)) => {},
        Tuple(.., Some(2)) => {},
        Tuple(.., None) => {},
        _ => {},
    }
    let t @ Tuple(_, two) = tuple;
}

fn ok_tuple(tuple: (Option<i32>, Option<i32>)) {
    if let (None, two) = tuple {}
    if let (one, Some(1)) = tuple {}
    match tuple {
        (Some(1), _) => {},
        (_, Some(1)) => {},
        _ => {},
    }
    match tuple {
        (.., Some(1)) => {},
        (.., Some(2)) => {},
        (.., None) => {},
        _ => {},
    }
    match tuple {
        (Some(1), ..) => {},
        (.., Some(1)) => {},
        _ => {},
    }
    let t @ (one, two) = tuple;
}

fn ok_array(array: [i32; 555]) {
    if let [.., 209, _] = array {}
}

fn ok_slice(slice: &[i32]) {
    // slices may panic if indexed, but will never panic
    // in a pattern since they're forced to check if they're empty
    match slice {
        [.., x] => {},
        [] => {},
    }
    if let [x, ..] = slice {}
}

enum Enum {
    Enum(Option<i32>, Option<i32>),
    OtherEnum,
}

fn ok_enum(enum1: Enum) {
    // Should never touch enums for this lint
    // since there's no way to access a field directly
    if let Enum::Enum(one, _) = enum1 {}
    match enum1 {
        Enum::Enum(one, ..) => {},
        Enum::Enum(_, _) => {},
        Enum::OtherEnum => {},
    }
}

fn main() {}
