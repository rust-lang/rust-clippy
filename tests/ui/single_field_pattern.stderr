error: this single-variant pattern matches no fields
  --> $DIR/single_field_pattern.rs:8:5
   |
LL |     let Struct { .. } = struct1;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::single-field-patterns` implied by `-D warnings`

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:9:5
   |
LL |     let Struct { field1, .. } = struct1;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     let field1 = struct1.field1;
   |         ~~~~~~   ~~~~~~~~~~~~~~

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:10:5
   |
LL |     let Struct { field1, field2: _ } = struct1;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     let field1 = struct1.field1;
   |         ~~~~~~   ~~~~~~~~~~~~~~

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:11:5
   |
LL | /     match struct1 {
LL | |         Struct { field1: Some(n), .. } if n >= 50 => {},
LL | |         Struct { field1: None, .. } => {},
LL | |         _ => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match struct1.field1 {
LL ~         Some(n) if n >= 50 => {},
LL ~         None => {},
LL ~         _ => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:16:5
   |
LL | /     match struct1 {
LL | |         Struct { field1: Some(n), .. } if n >= 50 => {},
LL | |         Struct { .. } => {},
LL | |         _ => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match struct1.field1 {
LL ~         Some(n) if n >= 50 => {},
LL ~         _ => {},
LL ~         _ => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:21:5
   |
LL | /     match struct1 {
LL | |         Struct { field1: Some(1), .. } => {},
LL | |         Struct { field1: Some(2), .. } => {},
LL | |         Struct { field1: None, .. } => {},
LL | |         _ => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match struct1.field1 {
LL ~         Some(1) => {},
LL ~         Some(2) => {},
LL ~         None => {},
LL ~         _ => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:27:5
   |
LL | /     match struct1 {
LL | |         Struct {
LL | |             field1: Some(_) | None, ..
LL | |         } => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match struct1.field1 {
LL ~         Some(_) | None => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:32:5
   |
LL |     while let Struct { field1: Some(5), .. } = struct1 {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     while let Some(5) = struct1.field1 {}
   |               ~~~~~~~   ~~~~~~~~~~~~~~

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:38:5
   |
LL |     if let Tuple(Some(x), _) = tuple {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     if let Some(x) = tuple.0 {}
   |            ~~~~~~~   ~~~~~~~

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:39:5
   |
LL |     if let Tuple(_, Some(y)) = tuple {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     if let Some(y) = tuple.1 {}
   |            ~~~~~~~   ~~~~~~~

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:40:5
   |
LL | /     match tuple {
LL | |         Tuple(Some(1), ..) => {},
LL | |         Tuple(Some(2), ..) => {},
LL | |         Tuple(a, ..) => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match tuple.0 {
LL ~         Some(1) => {},
LL ~         Some(2) => {},
LL ~         a => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:45:5
   |
LL | /     match tuple {
LL | |         Tuple(Some(1) | Some(42) | Some(6082), ..) => {},
LL | |         Tuple(a, ..) => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match tuple.0 {
LL ~         Some(1) | Some(42) | Some(6082) => {},
LL ~         a => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:52:5
   |
LL |     if let (Some(z), _) = tuple {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     if let Some(z) = tuple.0 {}
   |            ~~~~~~~   ~~~~~~~

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:53:5
   |
LL |     if let (_, Some(n)) = tuple {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     if let Some(n) = tuple.1 {}
   |            ~~~~~~~   ~~~~~~~

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:54:5
   |
LL | /     match tuple {
LL | |         (Some(1), ..) => {},
LL | |         (Some(2), ..) => {},
LL | |         (a, ..) => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match tuple.0 {
LL ~         Some(1) => {},
LL ~         Some(2) => {},
LL ~         a => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:59:5
   |
LL | /     match tuple {
LL | |         (Some(1) | Some(42) | Some(6082), ..) => {},
LL | |         (a, ..) => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match tuple.0 {
LL ~         Some(1) | Some(42) | Some(6082) => {},
LL ~         a => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:66:5
   |
LL | /     match array {
LL | |         [1 | 2, ..] => {},
LL | |         [x @ 3, ..] => {},
LL | |         [r @ 20..=65, ..] => {},
LL | |         [e, ..] => {},
LL | |     }
   | |_____^
   |
help: try this
   |
LL ~     match array[0] {
LL ~         1 | 2 => {},
LL ~         x @ 3 => {},
LL ~         r @ 20..=65 => {},
LL ~         e => {},
   |

error: this single-variant pattern only matches one field
  --> $DIR/single_field_pattern.rs:72:5
   |
LL |     if let [5, ..] = array {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: try this
   |
LL |     if let 5 = array[0] {}
   |            ~   ~~~~~~~~

error: aborting due to 18 previous errors

